### PR TITLE
Check for App Management permissions before updating apps.

### LINKED ISSUE #15484: brew upgrade for a cask creates a 0-byte file in Applications directory
### `brew doctor` output

```shell
Your system is ready to brew
```

### Verification

- [X] My "`brew doctor` output" above says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` twice and am still able to reproduce my issue.
- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### `brew config` output

```shell
HOMEBREW_VERSION: 4.0.19-24-ge96fd24
ORIGIN: https://github.com/Homebrew/brew
HEAD: e96fd24a7328a4913de2f7c56fd2c97d9aaa4572
Last commit: 5 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 34b732b38991b8b2f26b592d87287ebc77937f19
Core tap last commit: 4 months ago
Core tap branch: master
Core tap JSON: 25 May 15:52 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.40.1 => /opt/homebrew/bin/git
Curl: 7.87.0 => /usr/bin/curl
macOS: 13.3.1-arm64
CLT: 14.3.0.0.1.1679647830
Xcode: 14.3
Rosetta 2: false
```

### What were you trying to do (and why)?

Upgrade a cask

### What happened (include all command output)?

==> Upgrading 4 outdated packages:
raycast 1.51.3 -> 1.52.0
==> Upgrading raycast
==> Downloading https://releases.raycast.com/releases/1.52.0/download?build=universal
==> Downloading from https://d3jdrrl94b667u.cloudfront.net/Raycast_v1.52.0_71264205ebdf6061183dbe7343efa9de279b19fe_un
############################################################################################################### 100.0%
==> Quitting application 'com.raycast.macos'...
Application 'com.raycast.macos' quit successfully.
==> Backing App 'Raycast.app' up to '/opt/homebrew/Caskroom/raycast/1.51.3/Raycast.app'
==> Removing App '/Applications/Raycast.app'
==> Moving App 'Raycast.app' to '/Applications/Raycast.app'
cp: /Applications/Raycast.app/Contents: Operation not permitted
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents: unable to copy extended attributes to /Applications/Raycast.app/Contents: Operation not permitted
cp: /Applications/Raycast.app/Contents/_CodeSignature: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/_CodeSignature: unable to copy extended attributes to /Applications/Raycast.app/Contents/_CodeSignature: No such file or directory
cp: /Applications/Raycast.app/Contents/_CodeSignature/CodeResources: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents/_CodeSignature: No such file or directory
cp: chown: /Applications/Raycast.app/Contents/_CodeSignature: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents/_CodeSignature: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents/_CodeSignature: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/_CodeSignature: unable to copy ACL to /Applications/Raycast.app/Contents/_CodeSignature: No such file or directory
cp: /Applications/Raycast.app/Contents/MacOS: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/MacOS: unable to copy extended attributes to /Applications/Raycast.app/Contents/MacOS: No such file or directory
cp: /Applications/Raycast.app/Contents/MacOS/Raycast Restart Helper: No such file or directory
cp: /Applications/Raycast.app/Contents/MacOS/Raycast App Uninstall Helper: No such file or directory
cp: /Applications/Raycast.app/Contents/MacOS/Raycast: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents/MacOS: No such file or directory
cp: chown: /Applications/Raycast.app/Contents/MacOS: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents/MacOS: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents/MacOS: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/MacOS: unable to copy ACL to /Applications/Raycast.app/Contents/MacOS: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/Inter-Bold.otf: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/Inter-Regular.otf: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/emoji.json: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/JetBrainsMono-Bold.ttf: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/JetBrainsMono-Italic.ttf: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/AppIcon-Release.icns: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/Inter-SemiBoldItalic.otf: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/JetBrainsMono-Regular.ttf: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/package.json: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/.prettierrc: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/tsconfig.json: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/assets: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/assets: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/assets: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/assets/icon.png: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/assets: No such file or directory
cp: chown: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/assets: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/assets: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/assets: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/assets: unable to copy ACL to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/assets: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/.eslintrc.json: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/src: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/src: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/src: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/src/index.tsx: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/src: No such file or directory
cp: chown: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/src: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/src: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/src: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/src: unable to copy ACL to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started/src: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started: No such file or directory
cp: chown: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started: unable to copy ACL to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/getting-started: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit/.gitignore: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit: No such file or directory
cp: chown: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit: unable to copy ACL to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/teams-starter-kit: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/bin: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/bin: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/bin: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/bin/ray: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/bin: No such file or directory
cp: chown: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/bin: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/bin: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/bin: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/bin: unable to copy ACL to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/bin: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/hash: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/server: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/server: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/server: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/server/index.js: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/server: No such file or directory
cp: chown: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/server: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/server: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/server: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/server: unable to copy ACL to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/server: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/api: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/api: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/api: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/api/index.js: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/api: No such file or directory
cp: chown: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/api: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/api: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/api: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/api: unable to copy ACL to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/api: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi: unable to copy extended attributes to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi/raycast_napi.darwin-x64.node: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi/index.js: No such file or directory
cp: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi/raycast_napi.darwin-arm64.node: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi: No such file or directory
cp: chown: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.52.0/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi: unable to copy ACL to /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast/napi: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast: No such file or directory
cp: chown: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents/Resources/RaycastCommands_RaycastCommands.bundle/Contents/Resources/api/node_modules/@raycast: No such file or directory
cp: 
[SNIPPED BECAUSE OUTPUT WAS TOO LONG FOR GITHUB TO ACCEPT THE REQUEST]

cp: /Applications/Raycast.app/Contents/Info.plist: No such file or directory
cp: /Applications/Raycast.app/Contents/PkgInfo: No such file or directory
cp: utimensat: /Applications/Raycast.app/Contents: No such file or directory
cp: chown: /Applications/Raycast.app/Contents: No such file or directory
cp: chmod: /Applications/Raycast.app/Contents: No such file or directory
cp: chflags: /Applications/Raycast.app/Contents: No such file or directory
cp: /opt/homebrew/Caskroom/raycast/1.51.3/Raycast.app/Contents: unable to copy ACL to /Applications/Raycast.app/Contents: Operation not permitted

### What did you expect to happen?

Process should upgrade the application and replace it. The only solution is to delete the 0-byte application and then run `brew reinstall raycast` to create the new version of the application. I have also seen the same behavior with the following casks:

- obsidian
- stats
- visual-studio-code

I thought this might be a temporary bug that would clear itself, but I have been seeing this happen for about a month for these casks. All other casks seem to upgrade fine.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
`brew upgrade raycast`
```

### PR DESCRIPTION
**Pull request by JBYoshi**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

macOS Ventura added [new restrictions on apps being allowed to update other apps](https://developer.apple.com/videos/play/wwdc2022/10096/). In order for one app to update another, at least one of the following must be true:
* Both apps must be signed by the same developer
* The app being updated needs to explicitly allow the updater app
* The updater app needs to be granted "Full Disk Access" permission
* The updater app needs to be granted a new "App Management" permission

This didn't show up until #15138 was merged because prior to that, macOS considered Homebrew's app updates to be uninstalls and reinstalls, which discarded app data. Since #15138, macOS now recognizes that Homebrew is updating apps, so anyone who hasn't already granted the proper permissions to their terminal will get errors while copying the app files. This PR adds a check before app updates to ensure that the terminal has been granted permissions. If the terminal hasn't been granted permissions, Homebrew will exit and give the user instructions to grant their terminal the necessary permissions.

When setting up unit tests for this, I also moved the existing "upgrade properly handles non-writable directories" to a new group "when the directory is owned by root" to clarify the meaning and to better group the tests.

Fixes Homebrew/homebrew-cask#147383
Fixes Homebrew/homebrew-cask#147789
Fixes #15484

### CODE REVIEW

**Comment by SMillerDev** [comment]
There needs to be a way to use the old system because I'm not sure the Homebrew users in my company will be able to grant that permission.

**Comment by JBYoshi** [comment]
@ywwry66 I just tried that on RStudio and you seem to be right. I'm not sure why. Apple hasn't really documented the feature beyond their WWDC talk.

@SMillerDev @razvanazamfirei I can add an environment variable for that.

**Comment by SMillerDev** [comment]
Could just default to the old way if the check for permissions fails right?

**Comment by MikeMcQuaid** [comment]
> Could just default to the old way if the check for permissions fails right?

Yes, this would be better than an environment variable 👍🏻

**Comment by JBYoshi** [comment]
The reason I was considering an environment variable is because macOS doesn't let apps prompt for the App Management permission and then wait for the user to accept or deny it (as it does for permissions like the microphone); the permission is just automatically denied by default. If we ignore failures, first-time Homebrew users won't have a chance to grant those permissions before the first upgrade happens.

An alternative option would be to just print a warning message during the upgrade if the permission check fails. While the first upgrade would be done using the old behavior, the user could notice the warning and take action to grant the permissions for next time. Would that work?

**Comment by MikeMcQuaid** [comment]
> An alternative option would be to just print a warning message during the upgrade if the permission check fails. While the first upgrade would be done using the old behavior, the user could notice the warning and take action to grant the permissions for next time. Would that work?

Yeh, this makes sense 👍🏻

**Comment by ywwry66** [comment]
What about printing a message somewhere in `install.sh` as well?

**Comment by JBYoshi** [comment]
> What about printing a message somewhere in `install.sh` as well?

That would be nice, but I don't think we have a way to reliably do this until after someone's installed a cask. The terminal also doesn't appear in the App Management page until someone tries to modify an app that isn't one of thne built-in apps (which are under the System partition and can't be modified even with Full Disk Access). The best we could do is try it with an app like Pages that comes preinstalled but is managed by the App Store, and that would still fail if someone happened to have uninstalled the app in question.

**Comment by SMillerDev** [comment]
> That would be nice, but I don't think we have a way to reliably do this until after someone's installed a cask.

Additionally, people are already not reading the existing ones. Adding more won't improve anything

**Comment by JBYoshi** [comment]
By the way, I did push an update to switch to a warning message. Here's the message I've set up for now:

> Your terminal does not have App Management permissions, so Homebrew will delete and reinstall the app. This may result in some configurations (like notification settings or location in the Dock/Launchpad) being lost. To fix this, go to Settings > Security and Privacy > App Management and turn on the switch for your terminal.

**Comment by SMillerDev** [comment]
Sounds great, thanks for all the work here @JBYoshi. One thing I'm thinking is that it might be good to add "why does my app move in launchpad" or something to the FAQ doc and give this permission as an answer.

**Comment by JBYoshi** [comment]
@SMillerDev I've added that information.

**Review by SMillerDev — APPROVED** [review]

**Review by bevanjkay — APPROVED** [review]

File: Library/Homebrew/cask/quarantine.rb [hunk 1]
```diff
@@ -188,5 +188,68 @@ def self.copy_xattrs(from, to)
...
+    # Ensures that Homebrew has permission to update apps on macOS Ventura.
+    # This may be granted either through the App Management toggle or the Full Disk Access toggle.
+    # The system will only show a prompt for App Management, so we ask the user to grant that.
+    sig { params(app: Pathname, command: T.class_of(SystemCommand)).returns(T::Boolean) }
+    def self.app_management_permissions_granted?(app:, command:)
+      return true unless app.directory?
+
+      # To get macOS to prompt the user for permissions, we need to actually attempt to
+      # modify a file in the app.
+      test_file = app / ".homebrew-write-test"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/quarantine.rb:201]:
```suggestion
      test_file = app/".homebrew-write-test"
```

File: Library/Homebrew/cask/quarantine.rb [hunk 2]
```diff
@@ -188,5 +188,68 @@ def self.copy_xattrs(from, to)
...
+          raise e unless e.stderr.include?("touch: #{test_file}: Operation not permitted")
+        end
+      end
+
+      opoo <<~EOF
+        Your terminal does not have App Management permissions, so Homebrew will delete and reinstall the app.
+        This may result in some configurations (like notification settings or location in the Dock/Launchpad) being lost.
+        To fix this, go to Settings > Security and Privacy > App Management and turn on the switch for your terminal.
+      EOF
+      false
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cask/quarantine.rb:252]:
```suggestion
      EOF

      false
```

File: docs/FAQ.md [hunk 3]
```diff
@@ -220,3 +220,9 @@ Or use the `--greedy` flag:
     brew upgrade --greedy
 
 Refer to the `upgrade` section of the [`brew` manual page](Manpage.md) for more details.
+
+## Why do my apps lose their Dock position / Launchpad position / permission settings when I run `brew upgrade`?
```
Comment (MikeMcQuaid) [inline docs/FAQ.md:224]:
```suggestion
## Why do my cask apps lose their Dock position / Launchpad position / permission settings when I run `brew upgrade`?
```

File: docs/FAQ.md [hunk 4]
```diff
@@ -220,3 +220,9 @@ Or use the `--greedy` flag:
...
+
+Homebrew has two possible strategies to update apps: uninstalling the old version and reinstalling the new one, or replacing the contents of the app with the new contents. With the uninstall/reinstall strategy, [macOS thinks the app is being deleted without any intent to reinstall it](https://github.com/Homebrew/brew/pull/15138), and so it removes some internal metadata for the old app, including where it appears in the Dock and Launchpad and which permissions it's been granted. The content replacement strategy works around this by treating the upgrade as an in-place upgrade. However, starting in macOS Ventura, these in-place upgrades are only allowed when the updater application (in this case, the terminal running Homebrew) has [certain permissions granted](https://github.com/Homebrew/brew/pull/15483). Either the "App Management" or "Full Disk Access" permission will suffice.
```
Comment (MikeMcQuaid) [inline docs/FAQ.md:226]:
```suggestion
Homebrew has two possible strategies to update cask apps: uninstalling the old version and reinstalling the new one, or replacing the contents of the app with the new contents. With the uninstall/reinstall strategy, [macOS thinks the app is being deleted without any intent to reinstall it](https://github.com/Homebrew/brew/pull/15138), and so it removes some internal metadata for the old app, including where it appears in the Dock and Launchpad and which permissions it's been granted. The content replacement strategy works around this by treating the upgrade as an in-place upgrade. However, starting in macOS Ventura, these in-place upgrades are only allowed when the updater application (in this case, the terminal running Homebrew) has [certain permissions granted](https://github.com/Homebrew/brew/pull/15483). Either the "App Management" or "Full Disk Access" permission will suffice.
```

**Review by MikeMcQuaid — APPROVED** [review]
Great work! A few tiny nits I've applied myself.

**Pull request merged**